### PR TITLE
Handle multiple spokes in a ZTP repo

### PR DESCRIPTION
**Pull request by cwilkers**
##### SUMMARY

In the case where a single ztp repo deploys multiple spoke clusters, this PR will allow you to define a list `czga_clusters_list` to inform the role.

This has the following effects:
 
* When more than one cluster exists, the GitOps Application deployed by the role targets the `''` Namespace instead of `czga_clusters_namespace`.
* Namespaces and pull secrets are created for each cluster.

##### ISSUE TYPE

- New or Enhanced Feature

##### Tests

- [x] TestBos2: acm-hub ztp-spoke-vm-4.17 - SUCCESS https://www.distributed-ci.io/jobs/32c8d896-8e86-43a9-ac4e-874f2d6710ec/jobStates / SUCCESS https://www.distributed-ci.io/jobs/9b16d1dd-e512-42f6-81a7-124d093e2087/jobStates

Test-Hint: no-check

### CODE REVIEW

File: roles/configure_ztp_gitops_apps/tasks/main.yaml [hunk 1]
```diff
@@ -67,7 +73,9 @@
       ansible.builtin.replace:
         path: "{{ temp_dir.path }}/ztp/argocd/deployment/clusters-app.yaml"
         regexp: "clusters-sub"
-        replace: "{{ czga_clusters_namespace }}"
+        replace: "'{{ item }}'"
+        after: "^kind: Application"
+      loop: "{{ (czga_clusters_list | length == 1) | ternary(czga_clusters_list, ['']) }}"
```
[the remark this answers is not in the document]
Comment (cwilkers) [reply roles/configure_ztp_gitops_apps/tasks/main.yaml:78]: Yes ...
I had loops on the brain when writing this.

File: roles/configure_ztp_gitops_apps/tasks/main.yaml [hunk 2]
```diff
@@ -208,11 +217,12 @@
           kind: Secret
           metadata:
             name: assisted-deployment-pull-secret
-            namespace: "{{ czga_clusters_namespace }}"
+            namespace: "{{ item }}"
```
[the remark this answers is not in the document]
Comment (cwilkers) [reply roles/configure_ztp_gitops_apps/tasks/main.yaml:220]: It's not maximally flexible, but that is how I have been using it.
I guess there could be a use case where blue/green or dev/prod might require different access.

[the remark this answers is not in the document]
Comment (ramperher) [reply roles/configure_ztp_gitops_apps/tasks/main.yaml:220]: Could we document that fact so that it's clear that the same secret is used for all clusters?

**Comment by ramperher** [comment]
recheck

File: roles/configure_ztp_gitops_apps/tasks/main.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (cwilkers) [reply roles/configure_ztp_gitops_apps/tasks/main.yaml]: Unlooped now.

**Review by ramperher — COMMENTED** [review]
Also, don't forget to squash your commits to have a single one.
I've proposed to keep this work in the radar to analyze how feasible could the its implementation on DCI automation. Meantime, it will only be available through the collections. Thanks for raising this!

File: roles/configure_ztp_gitops_apps/tasks/main.yaml [hunk 3]
```diff
@@ -67,7 +73,8 @@
       ansible.builtin.replace:
         path: "{{ temp_dir.path }}/ztp/argocd/deployment/clusters-app.yaml"
         regexp: "clusters-sub"
-        replace: "{{ czga_clusters_namespace }}"
+        replace: "{{ (czga_clusters_list | length == 1) | ternary(czga_clusters_list, ['']) }}"
```
[the remark this answers is not in the document]
Comment (cwilkers) [reply roles/configure_ztp_gitops_apps/tasks/main.yaml]: Oh, that was clumsy of me. Thankfully a squash commits will hide a multitude of flubs.

**Comment by ramperher** [comment]
recheck

**Comment by ramperher** [comment]
Code looks fine, let's run a full job in our side with DCI to confirm this code is not breaking our current workflow. @cwilkers , if you can give a small report about the tests you've done from your side to confirm your use case is working, it would be great, then we can decide on when to merge this. Thx!

File: roles/configure_ztp_gitops_apps/tasks/main.yaml [hunk 4]
```diff
@@ -67,7 +73,8 @@
       ansible.builtin.replace:
         path: "{{ temp_dir.path }}/ztp/argocd/deployment/clusters-app.yaml"
         regexp: "clusters-sub"
-        replace: "{{ czga_clusters_namespace }}"
+        replace: "{{ (czga_clusters_list | length == 1) | ternary(czga_clusters_list[0], ['']) }}"
+        after: "^kind: Application"
```
Comment (ramperher) [inline roles/configure_ztp_gitops_apps/tasks/main.yaml:77]: This doesn't look like to be working. From https://www.distributed-ci.io/jobs/b444e94f-c12f-4306-b766-037af78a1866/jobStates?sort=date&task=d79548e4-484e-485a-9124-332b5e255149:
```
ok: [jumphost -> jumphost] => {"changed": false, "msg": "Pattern for before/after params did not match the given file: ^kind: Application(?P<subsection>.*)"}
```
Is this `after` field really needed?

Comment (cwilkers) [reply roles/configure_ztp_gitops_apps/tasks/main.yaml]: It seems that the beginning of line caret was at fault here, I didn't specify to treat the manifest as a multi-line, so only the very beginning of the manifest could count as the start of line. Removing the caret works. I also realized while testing that the empty string still was represented as an array, so I fixed that as well.

Comment (ramperher) [reply roles/configure_ztp_gitops_apps/tasks/main.yaml]: Ack, let's see in the next tests if there are no issues in our side.

**Comment by ramperher** [comment]
recheck

**Comment by ramperher** [comment]
recheck

**Review by ramperher — APPROVED** [review]
LGTM, default setup is not affected by this change. @cwilkers , could you confirm you're fine with your internal tests for this feature before merging?

**Comment by cwilkers** [comment]
@ramperher I have had several good e2e tests with the modifications and can confirm everything is good on my end.

Thanks!

**Comment by cwilkers** [comment]
@ramperher Could we look at getting this in? I'm actively using it for my HCP pipeline
